ReactWindows/ReactNative/Views/View/ReactViewManager.cs

```diff
@@ -0,0 +1,270 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Windows.UI.Xaml;
+using Windows.Storage;
+using Windows.Storage.AccessCache;
+using Windows.ApplicationModel.DataTransfer;
+using ReactNative.UIManager.Annotations;
+using ReactNative.UIManager;
+using Newtonsoft.Json.Linq;
+using ReactNative.UIManager.Events;
+
+namespace ReactNative.Views.View
+{
+    /// <summary>
+    /// The UWP-specific RCTView implementation.
+    /// </summary>
+    public class ReactViewManager: BaseViewManager
+    {
+        private class DragDropEvent : Event
+        {
+            private readonly string _name;
+            private readonly JObject _data;
+
+            public override string EventName => _name;
+            public override bool CanCoalesce => false;
+
+            public DragDropEvent(int viewTag, string name, JObject data)
+                : base(viewTag)
+            {
+                _name = name;
+                _data = data;
+            }
+
+            public override void Dispatch(RCTEventEmitter eventEmitter)
+            {
+                eventEmitter.receiveEvent(ViewTag, EventName, new JObject
+                {
+                    { "target", ViewTag },
+                    { "dataTransfer", _data }
+                });
+            }
+        }
+
+        /// <summary>
+        /// The exported custom bubbling event types.
+        /// </summary>
+        public override IReadOnlyDictionary<string, object> ExportedCustomBubblingEventTypeConstants
+        {
+            get
+            {
+                return new Dictionary<string, object>()
+                {
+                    {
+                        "topDragEnter",
+                        new Dictionary<string, object>()
+                        {
+                            {
+                                "phasedRegistrationNames",
+                                new Dictionary<string, string>()
+                                {
+                                    { "bubbled" , "onDragEnter" },
+                                    { "captured" , "onDragEnterCapture" }
+                                }
+                            }
+                        }
+                    },
+                    {
+                        "topDragOver",
+                        new Dictionary<string, object>()
+                        {
+                            {
+                                "phasedRegistrationNames",
+                                new Dictionary<string, string>()
+                                {
+                                    { "bubbled" , "onDragOver" },
+                                    { "captured" , "onDragOverCapture" }
+                                }
+                            }
+                        }
+                    },
+                    {
+                        "topDrop",
```

Comment:
>topDrop [](start = 25, length = 7)

Drop is a routed event, we'll have to be careful as there is the native bubbling mechanism that this could compete with. We should probably have some guidance for Windows whether routed events should be used as direct events only (and rely on native bubbling behaviors), or otherwise make sure that routed events that are bubbling in JS are always marked as "Handled".

Reply:
That's a good point. I did a small prototype with 3 nested views:

```jsx
        <View style={styles.view1} allowDrop={true} onDrop={e => log('view1', e)}>
          <View style={styles.view2} allowDrop={true} onDrop={e => log('view2', e)}>
            <View style={styles.view3} allowDrop={true} onDrop={e => log('view3', e)}>

            </View>
          </View>
        </View>
```

Then I dropped a file onto the inner view and got this:

```
XAML OnDrop ReactTag=8
XAML OnDrop ReactTag=9
XAML OnDrop ReactTag=10
[JS Info] view3 target=8
[JS Info] view2 target=8
[JS Info] view1 target=8
[JS Info] view2 target=9
[JS Info] view1 target=9
[JS Info] view1 target=10
```

So the two event bubbling mechanisms - XAML's and RN's - do interfere and number of JS events is like `N^2`. Not nice. In JS it's common to use `stopPropagation` in such event handlers, so I think we should go with the direct mode: `Handled=true` for XAML and direct events for RN.

Reply:
So I'm making all the four JS events direct (i.e. not bubbling). As for XAML events: drag-drop and drag-over will be direct, while drag-enter and drag-leave will continue bubble up, so JS will always get enter->leave pairs (if they weren't bubbling, JS would get a few drag-enter events and only one drag-leave for the view with the highest z-index).

Reply:
Anton - this sounds like a general problem. Can you write up some guidance on how to handle bubbling vs. direct events? There may even be a simple helper we can write to help you manage this.

---
In reply to: [159568508](https://github.com/Microsoft/react-native-windows/pull/1549#discussion_r159568508) [](ancestors = 159568508)